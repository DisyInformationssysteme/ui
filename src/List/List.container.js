import React, { PropTypes } from 'react';
import { api } from 'react-cmf';
import { Map } from 'immutable';
import { List as Component } from 'react-talend-components';

import { statePropTypes, stateWillMount } from '../state';
import { getActionsProps } from '../actionAPI';

export const DEFAULT_STATE = new Map({
	displayMode: 'table',
	searchQuery: '',
	limit: 0,
	offset: 0,
	sortOn: 'name',
	sortAsc: true,
});

export function getItems(context, props) {
	return props.items.map(
		item => Object.assign({}, item, {
			actions: getActionsProps(
				context, props.actions.items, item
			),
		})
	);
}

class List extends React.Component {
	static displayName = 'Container(List)';
	static propTypes = {
		actions: PropTypes.arrayOf(PropTypes.string),
		list: PropTypes.shape({
			columns: PropTypes.array,
			titleProps: PropTypes.object,
		}),
		toolbar: PropTypes.shape({
			sort: PropTypes.object,
			filter: PropTypes.object,
		}),
		items: PropTypes.arrayOf(PropTypes.object).isRequired,
		...statePropTypes,
	};

	static contextTypes = {
		store: PropTypes.object,
		registry: PropTypes.object,
	};

	constructor(props) {
		super(props);
		this.onSelectSortBy = this.onSelectSortBy.bind(this);
		this.onFilter = this.onFilter.bind(this);
		this.onSelectDisplayMode = this.onSelectDisplayMode.bind(this);
	}

	componentWillMount() {
		stateWillMount(this.props);
	}

	onSelectSortBy(event, payload) {
		this.props.updateState({
			sortOn: payload.field,
			sortAsc: !payload.isDescending,
		});
	}

	onFilter(event, payload) {
		this.props.updateState({ searchQuery: event });
	}

	onSelectDisplayMode(event, payload) {
		this.props.updateState({ displayMode: payload });
	}

	render() {
		const state = (this.props.state || DEFAULT_STATE).toJS();

		// list
		const items = getItems(this.context, this.props);
		const titleProps = this.props.list.titleProps;
		titleProps.onClick = (e, p) => {
			this.props.dispatch(
				api.action.getActionCreatorFunction(
					this.context,
					this.props.actions.title
				)(p.id)
			);
		};

		// toolbar
		const sort = this.props.toolbar.sort;
		sort.isDescending = !state.sortAsc;
		sort.field = state.sortOn;
		sort.onChange = (e, p) => {
			this.onSelectSortBy(e, p);
		};

		const filter = this.props.toolbar.filter;
		filter.onFilter = (e, p) => {
			this.onFilter(e, p);
		};

		const actionBar = {};
		const actions = this.props.actions;
		if (actions) {
			if (actions.left) {
				actionBar.left = getActionsProps(actions.left);
			}
			if (actions.right) {
				actionBar.right = getActionsProps(actions.right);
			}
		}

		const props = {
			displayMode: state.displayMode,
			list: {
				items,
				columns: this.props.list.columns,
				titleProps,
			},
			toolbar: {
				sort,
				filter,
				display: {
					onChange: (e, p) => {
						this.onSelectDisplayMode(e, p);
					},
				},
				//actionBar,
			},
		};
		return (<Component {...props} />);
	}
}

export default List;
