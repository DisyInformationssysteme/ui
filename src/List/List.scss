// List styles

.appBar{
  background-color: $tlnd-app-toolbar-bg-color;
  padding: 20px;
}
.searchInputContainer{
  width: 200px;
  :global(.form-control) {
    background-color: $tlnd-app-toolbar-bg-color;
  }
}
.container {

}
.bar {

}
.content {
  flex: 1;
  overflowY: 'auto';
  padding: 20px;
}
