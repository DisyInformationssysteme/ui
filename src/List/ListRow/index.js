import ListRow from './ListRow.component';

export default ListRow;
