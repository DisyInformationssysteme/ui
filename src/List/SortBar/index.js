import SortBar from './SortBar.component';

export default SortBar;
