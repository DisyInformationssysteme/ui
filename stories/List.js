import React from 'react';
import { storiesOf, action } from '@kadira/storybook';

import { List, IconsProvider } from '../src/index';

const props = {
	displayMode: 'table',
	list: {
		columns: [
			{ key: 'id', label: 'Id' },
			{ key: 'name', label: 'Name' },
			{ key: 'author', label: 'Author' },
			{ key: 'created', label: 'Created' },
			{ key: 'modified', label: 'Modified' },
		],
		items: [
			{
				id: 1,
				name: 'Hello world',
				created: '2016-09-22',
				modified: '2016-09-22',
				author: 'Jean-Pierre DUPONT',
				actions: [{
					label: 'edit',
					icon: 'talend-edit',
					onClick: action('onEdit'),
				}, {
					label: 'delete',
					icon: 'talend-delete',
					onClick: action('onDelete'),
				}, {
					displayMode: 'dropdown',
					label: 'related items',
					icon: 'talend-folder',
					items: [
						{
							label: 'document 1',
							onClick: action('document 1 click'),
						},
						{
							label: 'document 2',
							onClick: action('document 2 click'),
						},
					],
				}],
				icon: 'fa fa-file-excel-o',
			},
			{
				id: 2,
				name: 'Foo',
				created: '2016-09-22',
				modified: '2016-09-22',
				author: 'Jean-Pierre DUPONT',
				icon: 'fa fa-file-pdf-o',
			},
			{
				id: 2,
				name: 'Super long title to trigger overflow on tile rendering',
				created: '2016-09-22',
				modified: '2016-09-22',
				author: 'Jean-Pierre DUPONT with super long name',
			},
		],
		titleKey: 'name',
		iconKey: 'icon',
		onTitleClick: action('onClick'),
		onElementSelect: action('onSelect'),
	},
	toolbar: {
		onClickAdd: action('onClickAdd'),
		listActions: [
			{
				label: 'Delete selection',
				icon: 'talend-delete',
				onClick: action('delete'),
			},
		],
		onFilter: action('onFilter'),
		onSelectSortBy: action('onSelectSortBy'),
		onSelectDisplayMode: action('onSelectDisplayMode'),
		sortBy: [
			{ id: 'id', name: 'Id' },
			{ id: 'name', name: 'Name', selected: true },
		],
	},
};

storiesOf('List', module)
	.add('default', () => (
		<div>
			<h1>List</h1>
			<h2>Definition</h2>
			<p>Display a list by defining your.</p>
			<h2>Examples</h2>
			<IconsProvider />
			<List {...props} />
		</div>
	))
	.add('large', () => {
		const eprops = Object.assign({}, props);
		eprops.displayMode = 'large';
		return (
			<div>
				<h1>List</h1>
				<p>Display the list in large mode</p>
				<IconsProvider />
				<List {...eprops} />
			</div>
		);
	})
	.add('tile', () => {
		const tprops = Object.assign({}, props);
		tprops.displayMode = 'tile';
		return (
			<div>
				<h1>List</h1>
				<p>Display the list in tile mode</p>
				<IconsProvider />
				<List {...tprops} />
			</div>
		);
	});
