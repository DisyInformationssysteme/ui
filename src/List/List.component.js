import React from 'react';
import classNames from 'classnames';
import { Icon, actions } from 'react-cmf';

import ButtonsDispatcher from '../ButtonsDispatcher';
import SortBar from './SortBar';
import theme from './List.scss';

/**
 * @param {object} props react props
 * @example
<List name="Hello world"></List>
 */
function List(props) {
	const searchCSS = classNames(theme.searchInputContainer, 'form-group has-feedback pull-right');
	return (
		<div className={theme.container}>
			<div className={theme.appBar}>
				<div className={searchCSS}>
					<input
						type="search"
						name="search"
						value={this.props.searchQuery}
						onChange={this.onSearchQueryChange}
						className="form-control"
						placeholder={this.props.settings.searchLabel || 'search'}
					/>
					<Icon
						name="icon-search"
						className="form-control-feedback"
						aria-hidden="true"
					/>
				</div>
				<ButtonsDispatcher
					contentType={this.props.settings.contentType}
					category="list"
					icon btn="add"
				/>
			</div>
			<div className={theme.content}>
				<SortBar
					vocabulary={['name', 'step']}
					on={'name'}
					onChange={this.onSortChange}
				/>
				{this.props.items.map((item, i) => (
					<SortableListRow
						columns={this.props.settings.columns}
						contentType={this.props.settings.contentType}
						item={item} key={i}
					/>
				))}
			</div>
		</div>
	);
}

List.propTypes = {
	name: React.PropTypes.string,
};

export default List;
