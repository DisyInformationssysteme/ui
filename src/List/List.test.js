import React from 'react';
import renderer from 'react-test-renderer';

import List from './List.component';

jest.mock('react-dom');

const listProps = {
	columns: [
		{ key: 'id', label: 'Id' },
		{ key: 'name', label: 'Name' },
	],
	items: [
		{ id: 1, name: 'Hello world' },
		{ id: 2, name: 'Foo' },
		{ id: 3, name: 'Bar' },
	],
	titleProps: {
		onClick: jest.fn(),
	},
};

const toolbarProps = {
	onClickAdd: jest.fn(),
	onFilter: jest.fn(),
	onSelectSortBy: jest.fn(),
	onSelectDisplayMode: jest.fn(),
	sortOptions: [
		{ id: 'id', name: 'Id' },
		{ id: 'name', name: 'Name' },
	],
	sortBy: 'name',
};

const props = {
	displayMode: 'table',
	list: listProps,
	toolbar: toolbarProps,
};

describe('List', () => {
	it('should render', () => {
		const wrapper = renderer.create(<List {...props} />).toJSON();
		expect(wrapper).toMatchSnapshot();
	});

	it('should render id if provided', () => {
		const tProps = {
			id: 'context',
			...props,
		};
		const wrapper = renderer.create(<List {...tProps} />).toJSON();
		expect(wrapper).toMatchSnapshot();
	});

	it('should not render the toolbar without toolbar props', () => {
		const wrapper = renderer.create(<List displayMode="table" list={listProps} />).toJSON();
		expect(wrapper).toMatchSnapshot();
	});
});
