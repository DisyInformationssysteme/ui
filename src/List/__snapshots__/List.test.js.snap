exports[`List should render its name 1`] = `
<div
  className="tc-list">
  <div
    className="navbar navbar-default"
    role="toolbar">
    <div
      className="container-fluid">
      <ul
        className="nav navbar-nav"
        role={null}>
        <label
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-text"
          onSelect={null}>
          Display:
        </label>
        <li
          className="dropdown"
          style={undefined}>
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle"
            href=""
            id="grid"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button">
            <i
              aria-hidden="true"
              className="fa fa-bars"
              title={null} />
             
            <span
              className="caret" />
          </a>
          <ul
            aria-labelledby="grid"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-bars"
                  title={null} />
                 
                Table
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-list-alt"
                  title={null} />
                 
                Expanded
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                <i
                  aria-hidden="true"
                  className="fa fa-th-large"
                  title={null} />
                 
                Tile
              </a>
            </li>
          </ul>
        </li>
      </ul>
      <ul
        className="nav navbar-nav"
        role={null}>
        <label
          active={undefined}
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-text"
          onSelect={null}>
          Sort by:
        </label>
        <li
          className="dropdown"
          style={undefined}>
          <a
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-toggle"
            href=""
            id="tc-list-toolbar-sort-by"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button">
            Name
             
            <span
              className="caret" />
          </a>
          <ul
            aria-labelledby="tc-list-toolbar-sort-by"
            className="dropdown-menu"
            role="menu">
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Id
              </a>
            </li>
            <li
              className=""
              role="presentation"
              style={undefined}>
              <a
                href=""
                onClick={[Function]}
                onKeyDown={[Function]}
                role="menuitem"
                tabIndex="-1">
                Name
              </a>
            </li>
          </ul>
        </li>
        <button
          activeHref={undefined}
          activeKey={undefined}
          className="navbar-btn btn btn-link"
          disabled={false}
          onClick={[Function]}
          onSelect={null}
          role="button"
          type="button">
          <i
            aria-hidden="true"
            className="fa fa-sort-asc"
            title={null} />
          ASCENDING
        </button>
      </ul>
      <form
        className="navbar-form navbar-right"
        onSubmit={[Function]}
        role="search">
        <div
          className="form-group">
          <div
            className="input-group">
            <input
              aria-label="Filter"
              className="form-control"
              onChange={[Function]}
              placeholder="Filter"
              type="search" />
          </div>
          <button
            className="btn btn-link"
            disabled={false}
            title="Filter"
            type="submit">
            <i
              aria-hidden="true"
              className="fa fa-search" />
            <span
              className="sr-only">
              Filter
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
  <table
    className="table tc-list-display-table">
    <thead>
      <tr>
        <th>
          Id
        </th>
        <th>
          Name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            Hello world
          </button>
          <div
            className="tc-actions btn-group" />
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            Foo
          </button>
          <div
            className="tc-actions btn-group" />
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            role="link"
            type="button">
            Bar
          </button>
          <div
            className="tc-actions btn-group" />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
